fem/gridfunc.cpp

```diff
@@ -4550,4 +4550,409 @@ GridFunction *Extrude1DGridFunction(Mesh *mesh, Mesh *mesh2d,
+{
+   dest=&dest_;
+   FiniteElementSpace* space=dest->FESpace();
+   Mesh* mesh=space->GetMesh();
+   int dim=mesh->SpaceDimension();
+
+   if (dim==2) { kdt2D=new KDTree2D(); kdt3D=nullptr;}
+   if (dim==3) { kdt3D=new KDTree3D(); kdt2D=nullptr;}
+   std::vector<bool> indt; indt.resize(space->GetVSize()/space->GetVDim());
+   for (size_t i=0; i<indt.size(); i++) {indt[i]=true;}
+
+   minbb.SetSize(dim);
+   maxbb.SetSize(dim);
+
+   //set the loocal coordinates
+   {
+      ElementTransformation *trans;
+      const IntegrationRule* ir=nullptr;
+      Array<int> vdofs;
+      DenseMatrix elco;
+      int isca=1;
+      if (space->GetOrdering()==Ordering::byVDIM) {isca=space->GetVDim();}
+
+      //intialize the bounding box
+      const FiniteElement* el=space->GetFE(0);
+      trans = space->GetElementTransformation(0);
+      ir=&(el->GetNodes());
+      space->GetElementVDofs(0,vdofs);
+      elco.SetSize(dim,ir->GetNPoints());
+      trans->Transform(*ir,elco);
+      for (int d=0; d<dim; d++)
+      {
+         minbb[d]=elco(d,0);
+         maxbb[d]=elco(d,0);
+      }
+
+
+      for (int i=0; i<space->GetNE(); i++)
+      {
+         el=space->GetFE(i);
+         //get the element transformation
+         trans = space->GetElementTransformation(i);
+         ir=&(el->GetNodes());
+         space->GetElementVDofs(i,vdofs);
+         elco.SetSize(dim,ir->GetNPoints());
+         trans->Transform(*ir,elco);
+
+         if (dim==2)
+         {
+            for (int p=0; p<ir->GetNPoints(); p++)
+            {
+               int bind=vdofs[p]/isca;
+               if (indt[bind]==true)
+               {
+                  kdt2D->AddPoint(elco(0,p),elco(1,p),bind);
+                  indt[bind]=false;
+
+                  for (int d=0; d<2; d++)
+                  {
+                     if (minbb[d]>elco(d,p)) {minbb[d]=elco(d,p);}
+                     if (maxbb[d]<elco(d,p)) {maxbb[d]=elco(d,p);}
+                  }
+               }
+            }
+         }
+         else if (dim==3)
+         {
+            for (int p=0; p<ir->GetNPoints(); p++)
+            {
+               int bind=vdofs[p]/isca;
+               if (indt[bind]==true)
+               {
+                  kdt3D->AddPoint(elco(0,p),elco(1,p),elco(2,p),bind);
+                  indt[bind]=false;
+
+                  for (int d=0; d<3; d++)
+                  {
+                     if (minbb[d]>elco(d,p)) {minbb[d]=elco(d,p);}
+                     if (maxbb[d]<elco(d,p)) {maxbb[d]=elco(d,p);}
+                  }
+
+               }
+            }
+         }
+      }
+   }
+
+   // build the KDTree
+   if (dim==2) { kdt2D->Sort();}
+   else {kdt3D->Sort();}
+
+   minbb-=1e-8;
+   maxbb+=1e-8;
+
+}
+
+void KDTreeNodalProjection::Project(Vector& coords, Vector& src,
+                                    int ordering,double lerr)
+{
+   int dim=dest->FESpace()->GetMesh()->SpaceDimension();
+   int vd=dest->VectorDim();
+   int np=src.Size()/vd;
```

Comment:
How about `const int vectDim` instead of `int vd` and `const int numPoints` instead of `int np`?

Reply:
this will add quite long expressions in the indexing later - that is why I prefer the short names. I added comments after the definitions.